api/src/services/graphql/index.ts

```diff
@@ -193,23 +199,90 @@ export class GraphQLService {
 	 * Generate the GraphQL schema. Pulls from the schema information generated by the get-schema util.
 	 */
 	async getSchema(): Promise<GraphQLSchema>;
-	async getSchema(type: 'schema'): Promise<GraphQLSchema>;
-	async getSchema(type: 'sdl'): Promise<GraphQLSchema | string>;
-	async getSchema(type: 'schema' | 'sdl' = 'schema'): Promise<GraphQLSchema | string> {
+	async getSchema(type: 'schema', attempt?: number): Promise<GraphQLSchema>;
+	async getSchema(type: 'sdl', attempt?: number): Promise<GraphQLSchema | string>;
+	async getSchema(type: 'schema' | 'sdl' = 'schema', attempt = 0): Promise<GraphQLSchema | string> {
+		const MAX_ATTEMPTS = 3;
+
 		const key = `${this.scope}_${type}_${this.accountability?.role}_${this.accountability?.user}`;
 
-		const cachedSchema = cache.get(key);
+		const cachedSchema = cache.gqlSchema.get(key);
 
 		if (cachedSchema) return cachedSchema;
 
+		if (attempt >= MAX_ATTEMPTS) {
+			throw new Error(`Failed to get GQL sanitized schema: hit infinite loop`);
+		}
+
+		const lock = useLock();
+		const bus = useBus();
+
+		const lockKey = 'gqlSanitizedSchemaCache--preparing';
+		const messageKey = 'gqlSanitizedSchemaCache--done';
+		const processId = await lock.increment(lockKey);
+
+		if (processId >= (env['CACHE_SCHEMA_MAX_ITERATIONS'] as number)) {
+			await lock.delete(lockKey);
+		}
+
+		const currentProcessShouldHandleOperation = processId === 1;
+
+		if (!cache.globalSanitizedSchema) {
+			if (currentProcessShouldHandleOperation === true) {
+				try {
+					cache.globalSanitizedSchema = sanitizeGraphqlSchema(this.schema);
+				} finally {
+					await bus.publish(messageKey, { schema: cache.globalSanitizedSchema });
+					await lock.delete(lockKey);
+				}
+			} else {
+				logger.trace('GQL sanitized schema cache is prepared in another process, waiting for result.');
+
+				const timeout: Promise<any> = new Promise((_, reject) =>
+					setTimeout(reject, env['CACHE_SCHEMA_SYNC_TIMEOUT'] as number),
+				);
+
+				const subscription = new Promise<SchemaOverview>((resolve, reject) => {
+					bus.subscribe(messageKey, busListener).catch(reject);
+
+					function busListener(options: { schema: SchemaOverview | null }) {
+						if (options.schema === null) {
+							return reject();
```

Comment:
I believe we should call cleanup here as well? If I understand correctly it will setup the listeners again when we call `getSchema` on L263 below, without cleanup it will leave handing  listeners.

Reply:
Yes we should! There's https://github.com/directus/directus/pull/24506 to address it for `get-schema`. Also patched this change via https://github.com/directus/directus/pull/24501/commits/815dd567dd95e208560680e4c222532e1d7d2e25. Thanks! 🫡